File: benchmarks/call-iface-poly4/bench.go
package main

import "fmt"

type Adder interface {
	Add(int) int
}

type Add1 struct{}
type Add3 struct{}
type Add5 struct{}
type Add7 struct{}

func (a *Add1) Add(x int) int { return x + 1 }
func (a *Add3) Add(x int) int { return x + 3 }
func (a *Add5) Add(x int) int { return x + 5 }
func (a *Add7) Add(x int) int { return x + 7 }

func main() {
	iterations := 20000000
	var a0 Adder = &Add1{}
	var a1 Adder = &Add3{}
	var a2 Adder = &Add5{}
	var a3 Adder = &Add7{}
	acc := 0

	for i := 0; i < iterations; i++ {
		x := i & 1023
		slot := i & 3
		if slot == 0 {
			acc += a0.Add(x)
		} else if slot == 1 {
			acc += a1.Add(x)
		} else if slot == 2 {
			acc += a2.Add(x)
		} else {
			acc += a3.Add(x)
		}
	}

	fmt.Println(acc)
}

File: benchmarks/call-iface-poly4/bench.lua
local Add1 = {}
Add1.__index = Add1
function Add1.new() return setmetatable({}, Add1) end
function Add1:add(x) return x + 1 end

local Add3 = {}
Add3.__index = Add3
function Add3.new() return setmetatable({}, Add3) end
function Add3:add(x) return x + 3 end

local Add5 = {}
Add5.__index = Add5
function Add5.new() return setmetatable({}, Add5) end
function Add5:add(x) return x + 5 end

local Add7 = {}
Add7.__index = Add7
function Add7.new() return setmetatable({}, Add7) end
function Add7:add(x) return x + 7 end

local iterations = 20000000
local a0 = Add1.new()
local a1 = Add3.new()
local a2 = Add5.new()
local a3 = Add7.new()
local acc = 0

for i = 0, iterations - 1 do
  local x = i % 1024
  local slot = i % 4
  if slot == 0 then
    acc = acc + a0:add(x)
  elseif slot == 1 then
    acc = acc + a1:add(x)
  elseif slot == 2 then
    acc = acc + a2:add(x)
  else
    acc = acc + a3:add(x)
  end
end

print(acc)
